### PR TITLE
DM-1492: Save practice on publish

### PR DESCRIPTION
**Pull request by emyl3**
### JIRA issue link
[DM-1492](https://agile6.atlassian.net/secure/RapidBoard.jspa?rapidView=22&projectKey=DM&modal=detail&selectedIssue=DM-1492&quickFilter=31)

## Description - what does this code do?
This PR ensures that once the "Publish practice" button is clicked the practice is saved before it is published.

## Testing done - how did you test it/steps on how can another person can test it 
1. Log in as a user that can edit/publish a practice.
2. Ensure this user has access to an unpublished practice.
(If you do not have an unpublished practice, you create an unpublished practice as an admin in `.../admin/practices`)
3. Navigate to the unpublished practice
4. a. Make edits to the practice that do not fulfill conditions for publishing (e.g. do not fill out the contact support email)
4. b. Click "Publish practice". Edits to the practice should be saved but you should see a modal saying "Cannot publish"
4. c. BONUS: Make sure you don't see the "Publish practice" button on the "Adoptions" page.
5. a. Make edits to the practice and fulfill all conditions for publishing
5. b. You should be redirected to the practice page with an alert saying the practice is published. All edits should be saved
6. You should no longer see the "Publish practice" button when editing this practice
7. BONUS: Make sure the "Save your progress" and "Continue" buttons work as expected

## Screenshots, Gifs, Videos from application (if applicable)
N/A

## Link to mock-ups/mock ups (image file if you have it) (if applicable)
N/A

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Unit tests written (if applicable)
- [ ] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating JIRA issue
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/controllers/practices_controller.rb [hunk 1]
```diff
@@ -471,7 +387,8 @@ def practice_params
...
                                      additional_staffs_attributes: [:id, :_destroy, :title, :hours_per_week, :duration_in_weeks, :permanent],
                                      additional_resources_attributes: [:id, :_destroy, :name, :position, :description], required_staff_trainings_attributes: [:id, :_destroy, :title, :description], practice_creators_attributes: [:id, :_destroy, :name, :role, :avatar, :position, :delete_avatar, :crop_x, :crop_y, :crop_w, :crop_h],
-                                     publications_attributes: [:id, :_destroy, :title, :link, :position], additional_documents_attributes: [:id, :_destroy, :attachment, :title, :position], practice_permissions_attributes: [:id, :_destroy, :position, :name, :description])
+                                     publications_attributes: [:id, :_destroy, :title, :link, :position], additional_documents_attributes: [:id, :_destroy, :attachment, :title, :position], practice_permissions_attributes: [:id, :_destroy, :position, :name, :description],
+                                     practice_partner: {}, department: {})
```
Comment (emyl3) [inline app/controllers/practices_controller.rb:391]: ❓ is this too permissive?

**Comment by emyl3** [comment]
> When I click publish, my data does get saved, but there isn't a way for me to know until I refresh the page. If this is expected, please disregard.

Oh! Are you thinking it would be both helpful to display some sort of confirmation that the practice was saved both on successful (when you are redirected to the practice page and see the successful publish alert) and unsuccessful publish (e.g. when the "Cannot publish" modal appears?)
 🤔

**Comment by JoshingYou1** [comment]
> > When I click publish, my data does get saved, but there isn't a way for me to know until I refresh the page. If this is expected, please disregard.
> 
> @JoshingYou1 Are you thinking it would be both helpful to display some sort of confirmation that the practice was saved both on successful (when you are redirected to the practice page and see the successful publish alert) and unsuccessful publish (e.g. when the "Cannot publish" modal appears?)
> 🤔

Mainly the unsuccessful publish, though I'm unsure about the amount of effort that would need to go into it. Also, I didn't see a design for it, so we may not need it. Other than that, everything works as intended. Great work!

File: app/views/practices/_practice_editor_sidenav.html.erb [hunk 2]
```diff
@@ -62,12 +62,12 @@
         </ul>
         <% if params[:action] != 'adoptions' %>
           <div class="text-center">
-              <%= button_tag 'Save your progress', class: 'usa-button usa-button--outline width-full padding-y-2 margin-top-1', id: 'practice-editor-save-button' %>
+            <%= button_tag 'Save your progress', class: 'usa-button usa-button--outline width-full padding-y-2 margin-top-1', id: 'practice-editor-save-button' %>
           </div>
         <% end %>
-        <% unless @practice.published %>
+        <% if !@practice.published && params[:action] != 'adoptions' %>
```
Comment (ahay-agile6) [inline app/views/practices/_practice_editor_sidenav.html.erb:68]: i think we still want this on the adoptions page, no?

File: app/controllers/practices_controller.rb (hunk 1 above)
Comment (ahay-agile6) [reply app/controllers/practices_controller.rb:391]: what are these for?

File: app/services/save_practice_service.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    updated = @practice.update(@practice_params)
+
+    update_practice_partner_practices
+    update_department_practices
+    remove_attachments
+    manipulate_avatars
+    remove_main_display_image
+    crop_main_display_image
+
+    return updated
```
Comment (ahay-agile6) [inline app/services/save_practice_service.rb:21]: remove `return`

File: app/controllers/practices_controller.rb [hunk 4]
```diff
@@ -364,6 +277,9 @@ def published
   end
 
   def publication_validation
+    pr_params = { practice: @practice, practice_params: practice_params }
+    SavePracticeService.new(pr_params).save_practice
```
Comment (ahay-agile6) [inline app/controllers/practices_controller.rb:281]: if this returns `false` what happens?
do we still try to publish?
should we still try to publish? :thinking:

File: app/services/save_practice_service.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  def save_practice
+    updated = @practice.update(@practice_params)
+
+    update_practice_partner_practices
+    update_department_practices
+    remove_attachments
+    manipulate_avatars
+    remove_main_display_image
+    crop_main_display_image
```
Comment (ahay-agile6) [inline app/services/save_practice_service.rb:19]: if any of these fail, what happens?
perhaps we should wrap this in a `rescue` and return false if something happens and tell the user

File: app/views/practices/_practice_editor_sidenav.html.erb (hunk 2 above)
Comment (emyl3) [reply app/views/practices/_practice_editor_sidenav.html.erb:68]: The issue with publishing on the adoption page is that we attempt to `        Rails.fire(form[0], 'submit');` submit a `form` that doesn't exist on that page... Should I create a workaround to publish on the adoptions page or leave as-is?

File: app/controllers/practices_controller.rb (hunk 4 above)
Comment (emyl3) [reply app/controllers/practices_controller.rb:281]: Oh... 😓 Nothing happens when `false` is returned from saving the practice...
My inclination is if there's an issue with saving the practice, we shouldn't try to publish....
If they saved successfully and they meet the requirements for publishing, then they should be allowed to publish. Do you have any opinions on this?

File: app/services/save_practice_service.rb (hunk 5 above)
Comment (emyl3) [reply app/services/save_practice_service.rb:19]: that's a great idea 😅 will do!

File: app/controllers/practices_controller.rb (hunk 1 above)
Comment (emyl3) [reply app/controllers/practices_controller.rb:391]: ahh so the 'practice_partner` and `department` params can be passed to the `SavePracticeService`'s `save_practice` method and they can be updated as needed in the `update_practice_partner_practices` and `update_department_practices` methods... haha is there a better way to do this? 😅

File: app/controllers/practices_controller.rb (hunk 4 above)
Comment (ahay-agile6) [reply app/controllers/practices_controller.rb:281]: i agree

File: app/views/practices/_practice_editor_sidenav.html.erb (hunk 2 above)
Comment (ahay-agile6) [reply app/views/practices/_practice_editor_sidenav.html.erb:68]: yeah, we can probably use the original link_to we had on the adoptions page

File: app/controllers/practices_controller.rb [hunk 6]
```diff
@@ -372,12 +277,23 @@ def published
...
     respond_to do |format|
-      if @practice.name.present? && @practice.tagline.present? && @practice.initiating_facility.present? && @practice.date_initiated.present? && @practice.summary.present? && @practice.support_network_email.present?
-        @practice.update_attributes(published: true)
-        flash[:notice] = "#{@practice.name} has been successfully published to the Diffusion Marketplace"
-        # format.html { redirect_to practice_path(@practice), notice: flash[:notice] }
-        format.js { render js: "window.location='#{practice_path(@practice)}'" }
+      if can_publish
+        # if there is an error with updating the practice, alert the user
+        if updated.is_a?(StandardError)
+          flash[:error] = "There was an #{updated.message}. The practice was not saved or published."
```
Comment (emyl3) [inline app/controllers/practices_controller.rb:290]: ❓ Let me know if this copy can be improved...

File: app/services/save_practice_service.rb (hunk 3 above)
Comment (emyl3) [reply app/services/save_practice_service.rb]: removed!

**Comment by JoshingYou1** [comment]
Everything looks great! Good work Elisa!

**Pull request merged**